design-documents/stateless-api-token.md

```diff
@@ -0,0 +1,147 @@
+## Introduce stateless web API tokens
+&nbsp;&nbsp;&nbsp;&nbsp;Authentication tokens used to access web APIs (REST, SOAP, GraphQL) should contain enough
+information to authenticate a user without having to store session data on server side.
+ 
+This approach has following advantages:
+* In case of distributed deployment instances don't have to access single storage/service to authenticate a user
+* Saves storage space
+* Faster to decode/decrypt than load from storage
+* Don't have to implement token protection mechanism on backend to keep tokens from being stolen
+ 
+ 
+### Current situation
+&nbsp;&nbsp;&nbsp;&nbsp;Magento stores tokens in RDMS table named `oauth_token` which is originally meant to store long-term integration tokens
+but also used for admin/customer user tokens. Token rows also contain user type, issued data and expiration data which
+is retrieved when authenticating users for web API via token passed in the header. Admin and customer tokens may be
+manually expired by admin users by updating `oauth_token` table data. Tokens themselves are random strings of static
+length.
+ 
+With described above circumstances in order to perform authentication Magento instances have to have access to same DB
+and perform (relatively) expensive SQL queries to DB. Te more users Magento has the more tokens it has to store
+until they expire and run an extra cron job to remove expired ones.
+ 
+ 
+### Overview of new stateless tokens implementation
+* `TokenUserContext` to be updated to decode new tokens and use old logic for existing tokens
+* [Create and implement](#spi-for-token-data-manager) SPIs for providing user data to be encoded in tokens and reading it
+* Create and [implement](#encoding-and-signing-tokens) SPI for singing and encoding token data
+* Update token revoke functionality to use cache storage to store revoke request instead of DB for customer/admin tokens
+  [details](#how-to-handle-manual-expiration)
+ 
+ 
+### Encoding and signing tokens
+Tokens must have enough data encoded in order to authenticate a user. At minimum it would contain:
+* user type
+* user ID
+* issued timestamp
+* expires timestamp (in order to be able to validate token without knowing token lifespan)
+ 
+3rd-party developers will be able to extend this list via [the SPI](#spi-for-token-data-manager).
+ 
+After serialization this data must be signed to confirm origin. Magento instance(es) will have the same secret used
+to decode the tokens without having to validate them via DB.
+ 
+Magento will have _SPI_ that will handle encoding/decoding tokens and allow 3rd-party developers to change
+how tokens are handled:
+ 
+```php
+interface AuthData
+{
+    public function getIssued(): int;
+
+    public function getExpires(): int;
+
+    public function getUserData(): array;
+}
+
+
+interface TokenEncoderInterface
+{
+    public function encode(AuthData $data): string;
+
+    public function decode(string $token): AuthData;
```

Comment:
How the verification will look like?

Reply:
Verification will be done by an implementation of TokenEncoderInterface. Special exception will be thrown if it fails. As for the implementation itself - if we go with own implementation route then we'd just compare signatures by recreating them, otherwise we'll rely on a JWT library